### ISSUE TITLE
BUGFIX v0.5.0 - Handling of missing values in repeated measurements

### ISSUE THREAD

**Issue by raphaelvallat** [raphaelvallat/pingouin#206]
:bangbang: | This issue explains a critical bug that will be fixed in Pingouin v0.5.0. Please read carefully, and make sure to double check all your results with another statistical software.
:---: | :---

### One-way repeated measures ANOVA

Let's create a repeated measures dataset, in three different formats:

```python
import numpy as np
import pandas as pd
import pingouin as pg

# Create a wide-format dataframe, with missing values
df_a = pg.read_dataset("rm_anova_wide")
df_a['sub'] = np.arange(df_a.shape[0])
df_a['group'] = ['A'] * 5 + ['B'] * 7
df_a.set_index(['sub', 'group'], inplace=True)

# Convert to long-format
df_b = df_a.melt(var_name="time", ignore_index=False).sort_index().reset_index()

# Convert to long-format and remove rows with missing values
df_c = df_b.dropna().reset_index(drop=True)
```

![image](https://user-images.githubusercontent.com/25703792/138565129-25b5bfd5-9587-4b26-8fe9-1a1ed7a4e775.png)

Note how `df_c` has no explicit missing value. Instead, the missing values are "implicit", such that there are no rows in the dataframe when the data is missing. However, all these datasets contain exactly the same non-missing data. Therefore, any call to [pg.rm_anova](https://pingouin-stats.org/generated/pingouin.rm_anova.html) should lead to similar results. Unfortunately, this was not the case in versions of Pingouin <0.5.0. Indeed, `df_a` and `df_b` would return similar (correct) results, but `df_c` gave wrong results.

```python
print(pg.rm_anova(df_a))
print(df_b.rm_anova(dv="value", within="time", subject="sub"))
print(df_c.rm_anova(dv="value", within="time", subject="sub"))
```

![image](https://user-images.githubusercontent.com/25703792/138565307-947e6531-d247-4603-9084-9a26434fdb15.png)

Pay attention to the degrees of freedom (`ddof2`): `df_c` has more degrees of freedom, because no listwise deletion was applied. This leads to a smaller p-value. By contrast, in `df_a` and `df_b`, any subject/rows with missing values were completely removed before calculating the ANOVA. This is the default behavior of [JASP](https://jasp-stats.org/download/) and other statistical softwares, aka complete-case analysis. This has now been fixed in Pingouin 0.5.0, such that results would be similar for all three dataframes.

Importantly, this issue did not impact [pg.pairwise_ttests](https://pingouin-stats.org/generated/pingouin.pairwise_ttests.html). Indeed:

```python
df_b.pairwise_ttests(dv="value", within="time", subject="sub")
```
would lead to valid results (note how the degrees of freedom is 8 meaning that listwise deletion of rows with missing values was automatically performed)

![image](https://user-images.githubusercontent.com/25703792/138565650-c7c8abd2-ba48-4fa9-89ed-c2573729f95b.png)

However, calling `df_c` would lead to an error:

```python
df_c.pairwise_ttests(dv="value", within="time", subject="sub")
```

![image](https://user-images.githubusercontent.com/25703792/138565691-415af8c7-b435-4f5b-8636-f1de1540cca2.png)

Of note, we can also disable the automatic listwise deletion by using `nan_policy="pairwise"`, in which case the degrees of freedom are larger because missing values are only removed separately for each pairwise test:

![image](https://user-images.githubusercontent.com/25703792/138565753-482c15b7-24b2-4997-b869-55234aa55888.png)

****

### Mixed ANOVA

The same issue applies to [pg.mixed_anova](https://pingouin-stats.org/generated/pingouin.mixed_anova.html). Indeed,

```python
print(df_b.mixed_anova(dv="value", within="time", between="group", subject="sub"))
print(df_c.mixed_anova(dv="value", within="time", between="group", subject="sub"))
```

![image](https://user-images.githubusercontent.com/25703792/138566144-4cb62813-be2c-4ac0-9d71-7c8340742659.png)

Here again, `df_b` is the correct one. The degrees of freedom are smaller, meaning that a listwise deletion was applied: all participants with one or more missing value were entirely removed. This is also the default behavior in [JASP](https://jasp-stats.org/download/).

As earlier, this issue did not affect [pg.pairwise_ttests](https://pingouin-stats.org/generated/pingouin.pairwise_ttests.html), as it would return an error for `df_c`.

****

### Two-way repeated measures ANOVA

```python
# Long-format dataframe with explicit NaN
df_b = pg.read_dataset("rm_anova2").sort_values(by=['Time', 'Metric']).reset_index(drop=True)
df_b['Performance'] = df_b['Performance'].astype(float)
df_b.at[0, "Performance"] = np.nan

# Implicit missing value
df_c = df_b.copy().drop(index=0).reset_index(drop=True)

# Wide-format, explicit listwise deletion and convert to long-format
df_piv = df_b.pivot_table(index=['Subject'], columns=['Time', 'Metric'], values="Performance")
df_a = df_piv.dropna().melt(value_name="Performance", ignore_index=False).reset_index()
```

![image](https://user-images.githubusercontent.com/25703792/138571831-8b28dc1a-ee80-4b94-9e07-d5a885d13445.png)

This is the output that we get in [JASP](https://jasp-stats.org/download/)(when using `df_piv` but without .dropna()):

![image](https://user-images.githubusercontent.com/25703792/138571907-60bb3c0e-5039-41dc-b1e8-567779322be6.png)

No, if we run these same two-way repeated measures ANOVA in Pingouin <0.5.0, only `df_a` will produce the same output as JASP. This is because we have enforced a complete-case analysis with [pandas.dropna](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html). This is also the new behavior in Pingouin >0.5.0. You can check with these lines:

```python
df_a.rm_anova(dv="Performance", within=["Time", 'Metric'], subject="Subject")
df_b.rm_anova(dv="Performance", within=["Time", 'Metric'], subject="Subject")
df_c.rm_anova(dv="Performance", within=["Time", 'Metric'], subject="Subject")
```

****

**TL:DR**

In Pingouin <0.5.0, listwise deletion of subjects (or rows) with missing values was not strictly enforced in repeated measures or mixed ANOVA, depending on the input data format (if missing values were explicit or implicit). Pingouin 0.5.0 now uses a stricter complete-case analysis regardless of the input data format, which is the same behavior as JASP. Practically, this may lead to a decrease in the degrees of freedom and an increase in the p-values. We therefore highly recommend that you double check any results obtained with these functions. For future analyses, we also highly recommend that you manually deal with missing values (via imputation or listwise removal) before calling Pingouin's functions.
